src/os_posix/os_mtx_cond.c

```diff
@@ -103,27 +103,30 @@ void
 __wt_cond_signal(WT_SESSION_IMPL *session, WT_CONDVAR *cond)
 {
 	WT_DECL_RET;
-	bool locked;
-
-	locked = false;
 
 	__wt_verbose(session, WT_VERB_MUTEX, "signal %s", cond->name);
 
-	/* Fast path if already signalled. */
-	if (cond->waiters == -1)
-		return;
+	/*
+	 * Our callers are often setting flags to cause a thread to exit. Add
+	 * a barrier to ensure the flags are seen by the threads.
+	 */
+	WT_WRITE_BARRIER();
```

Comment:
I ran `medium-multi-btree-log.wtperf` and `evict-btree-stress-multi.wtperf` (since eviction and logging are heavy consumers of `__wt_cond_signal())`, and they're about the same as before. Let me know if you have other workloads you think might exercise this change.